src/irmin-pack/inode.ml

```diff
@@ -1263,8 +1273,37 @@ struct
         proof_of_concrete t.hash p
 
       let of_proof la ~depth (proof : proof) =
-        let c = concrete_of_proof ~depth proof in
-        match of_concrete ~depth la c with Ok v -> Some v | Error _ -> None
+        match proof with
+        | `Values vs when List.length vs > Conf.entries -> (
+            if depth <> 0 then None
+            else
+              (* [proof] is a big stable inode that was unshallowed and encoded
+                 in a [Values], it needs to be converted back to a [Tree]
+                 shallowed. *)
+              let t = of_seq Total (List.to_seq vs) in
+              (* Compute the hash right away (not lazily) *)
+              let hash = hash t in
+              let hash = lazy hash in
+              match t.v with
+              | Values _ -> assert false
+              | Tree { depth; length; entries } ->
+                  let entries =
+                    Array.map
+                      (function
+                        | None -> None
+                        | Some (Total_ptr ptr) ->
+                            let hash = Lazy.force ptr.hash in
+                            Some (Ptr.of_hash la hash))
```

Comment:
This will fail if `la = Total`, but this is fine since AFAICS we only call this with `Partial`. Maybe we should either add a comment or consume `~find` directly here to ensure we don't accidentally break this in future?

Reply:
I've changed `of_proof` to only work on `Partial` layouts